java/tools/src/main/java/org/apache/arrow/tools/EchoServer.java

```diff
@@ -132,7 +132,8 @@ public void run() throws IOException {
             }
           }
           writer.end();
-          Preconditions.checkState(reader.bytesRead() == writer.bytesWritten());
+          // reader does not read padding bytes.
```

Comment:
Hmm, this doesn't seem quite right. The padded bytes are included in the `messageLength`, then that length is read using `ReadChannel.readFully` so `reader.bytesRead()` should include the padded bytes as well

Reply:
Oh wait, is this the additional 4 bytes sent at the end of stream? What happens if the socket isn't closed and then another stream or other data is sent over it? I think that extra 4 bytes will still need to be read and cause problems for the next stream. 

@wesm @emkornfield , does C++ ignore the last 4-bytes at the EOS, since now it is an 8-byte value?

Reply:
Good catch, I looked again and I think you are right, the additional 4 bytes should be read.
In this case, may be we should also pass option to reader and read the additional 4 bytes if necessary?

Reply:
It looks like the C++ implementation only reads the first 4 bytes:

https://github.com/apache/arrow/blob/ARROW-6313-flatbuffer-alignment/cpp/src/arrow/ipc/message.cc#L327-L345

How else are we supposed to know if we should read 4 or 8 bytes? Maybe for the new format, an 8-byte EOS token should look like `{0xFFFFFFFF, 0x00000000}`, so we read the continutation token first, and then know to read the next 4 bytes, which are then `0` to signal EOS.

Reply:
Could the reader just remember the state, so if it reads the continuation
token from the beginning, then read all 8 bytes at the end?